### ISSUE TITLE
DL4J: Lack of input validation for batch norm?

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#6403]
As reported in gitter by @DrChainsaw 

> DrChainsaw @DrChainsaw Sep 09 01:35
@AlexDBlack Running on CPU immediately showed I didn't handle Batchnormalization correctly. Setting nIn on that layer does not do anything. After fixing this I no longer crash at the same point with CUDA so it seems like the CUDA backend does not catch this immediately. Exception when using CPU: https://gist.github.com/DrChainsaw/cc5a9d4dfd5cfc418cebe9e140bf469a
Btw, the fix was to set nOut to what I wanted nIn to be
I'm not sure I can put this in a "crisp" issue. It seems that some precondition which is checked with native backend is not checked with CUDA backend and this causes (or might cause) a delayed error when executing. I could feedforward a few times just fine with the CUDA backend with the "corrupt" batchnorm and nothing seemed off
